ra2ce/network/avg_speed/avg_speed_calculator.py

```diff
@@ -0,0 +1,215 @@
+            if _base_rt in _avg_speed.road_types:
+                return _avg_speed.get_avg_speed(_base_rt)
+            return 0.0
+
+        def get_uncombined_speed(combined_types: list[RoadTypeEnum]) -> float:
+            for _rt in combined_types:
+                if _rt in _avg_speed.road_types:
+                    return _avg_speed.get_avg_speed(_rt)
+            return 0.0
+
+        _avg_speed = AvgSpeed()
+
+        for _rt in list(
+            set(
+                str(edata[road_type_col_name])
+                for _, _, edata in self.graph.edges.data()
+            )
+        ):
+            edge_data = [
+                (self.parse_speed(edata["maxspeed"]), edata["length"])
+                for _, _, edata in self.graph.edges.data()
+                if (edata[road_type_col_name] == _rt) & ("maxspeed" in edata)
+            ]
+            if not edge_data:
+                _avg_speed.set_avg_speed(_rt, 0.0)
+            else:
+                # Calculate weighted average speed
+                edge_speed, edge_length = zip(*edge_data)
+                _avg_speed.set_avg_speed(
+                    _rt,
+                    sum(s * l for s, l in zip(edge_speed, edge_length))
+                    / sum(edge_length),
+                )
+
+        # For all types without an average speed, take one that is most likely.
+        _zero_rts = [
+            _rt for _rt in _avg_speed.road_types if _avg_speed.get_avg_speed(_rt) == 0
+        ]
+        if _zero_rts:
+            logging.info(
+                "Not all of the edges contain a 'maxspeed' attribute. RA2CE will guess the right average maximum "
+                "speed per road type that does not contain a 'maxspeed' attribute. Please check the average speed CSV to ensure correct speeds."
+            )
+            for _rt in _zero_rts:
+                if isinstance(_rt, list):
+                    # Try to get the speed from the uncombined road types in the list.
+                    _uncombined_speed = get_uncombined_speed(_rt)
+                    if _uncombined_speed:
+                        _avg_speed.set_avg_speed(_rt, _uncombined_speed)
+                        continue
+                # For the links take the one of the same type of the main roads.
+                if "_link" in str(_rt):
+                    _main_speed = get_main_speed(_rt)
+                    if _main_speed:
+                        _avg_speed.set_avg_speed(_rt, _main_speed)
+                        continue
+                _avg_speed.set_avg_speed(_rt, _avg_speed.default_speed)
+                logging.warning(
+                    "Default speed have been assigned to road type %s. Please check the average speed CSV, "
+                    "enter the right average speed for this road type and run RA2CE again.",
+                    _rt,
+                )
+
+        return _avg_speed
+
+    def calculate(self) -> AvgSpeed:
```

Comment:
For what I see we only use this method internally, so first step:
```suggestion
    def _calculate(self) -> AvgSpeed:
```
Second is, why wait to calculate until `asign` is called rather than during the initialization of the object? This way you don't require to store / define the `output_graph_dir` field. Also saving you the extra `if not self.avg_speed`

Reply:
I started off with the idea that `calculate` could also be called, after which the assign is called with a different `graph`. For now that's not needed, so I adjusted it according to your suggestions.